common/channel/src/message_queues.rs

```diff
@@ -0,0 +1,133 @@
+
+/// ValidatorMessage trait needs to be implemented by every message which gets pushed into the Libra
+/// channel so that we can ensure fairness among validators
+pub trait ValidatorMessage {
+    /// Extract the Validator from which this message arrived. This
+    /// will be used for ensuring fairness among validators.
+    fn get_validator(&self) -> AccountAddress;
+}
+
+/// QueueStyle is an enum which can be used as a configuration option for
+/// PerValidatorQueue. Since the queue per validator is going to be bounded,
+/// QueueStyle also determines the policy for dropping messages.
+/// With LIFO, oldest messages are dropped.
+/// With FIFO, newest messages are dropped.
+#[derive(Clone, Copy)]
+pub enum QueueStyle {
+    LIFO,
+    FIFO,
+}
+
+/// PerValidatorQueue maintains a queue of messages per validator. It
+/// is a bounded queue per validator and the style (FIFO, LIFO) is
+/// configurable. When a new message is added using `push`, it is added to
+/// the validator's queue.
+/// When `pop` is called, the next message is picked from one
+/// of the validator's queue and returned. This happens in a round-robin
+/// fashion among validators.
+/// If there are no messages, in any of the queues, `None` is returned.
+pub struct PerValidatorQueue<T> {
+    /// QueueStyle for the messages stored per validator
+    queue_style: QueueStyle,
+    /// per_validator_queue maintains a map from a Validator to a queue
+    /// of all the messages from that Validator. A Validator is
+    /// represented by AccountAddress
+    per_validator_queue: HashMap<AccountAddress, VecDeque<T>>,
+    /// This is a (round-robin)queue of Validators which have pending messages
+    /// This queue will be used for performing round robin among
+    /// Validators for choosing the next message
+    round_robin_queue: VecDeque<AccountAddress>,
+    /// Maximum number of messages to store per validator
+    max_queue_size: usize,
+}
+
+impl<T> PerValidatorQueue<T> {
+    /// Create a new PerValidatorQueue with the provided QueueStyle and
+    /// max_queue_size_per_validator
+    pub fn new(queue_style: QueueStyle, max_queue_size_per_validator: usize) -> Self {
+        assert!(
+            max_queue_size_per_validator > 0,
+            "max_queue_size_per_validator should be > 0"
+        );
+        Self {
+            queue_style,
+            max_queue_size: max_queue_size_per_validator,
+            per_validator_queue: HashMap::new(),
+            round_robin_queue: VecDeque::new(),
+        }
+    }
+
+    /// Given a validator, pops the message from its queue and returns the message
+    /// It also returns a boolean indicating whether the validators queue is empty
+    /// after popping the message
+    fn pop_from_validator_queue(&mut self, validator: AccountAddress) -> (Option<T>, bool) {
+        if let Some(q) = self.per_validator_queue.get_mut(&validator) {
+            if q.is_empty() {
+                return (None, true);
+            }
```

Comment:
I don't think you need it (`pop_front()` and `pop_back()` should return Option)

Reply:
Updated